File: src/lib/insurance-matching.ts
import { config } from "./config";

export type InsurancePlanType = "term" | "health";

export interface InsurancePlanReferenceLike {
  id: string;
  planType: InsurancePlanType;
  sumAssuredMax: number;
}

// Closest sum-assured match to the computed gap — never "largest" or
// "cheapest" — to avoid the selection itself acting as a ranking signal
// (see .agents/context/subsystem-notes.md).
export function selectInsuranceExamples<T extends InsurancePlanReferenceLike>(
  planType: InsurancePlanType,
  gapAmount: number,
  plans: T[],
  count: number = config.insurance_examples_per_gap_type,
): T[] {
  return plans
    .filter((plan) => plan.planType === planType)
    .sort(
      (a, b) =>
        Math.abs(a.sumAssuredMax - gapAmount) - Math.abs(b.sumAssuredMax - gapAmount),
    )
    .slice(0, count);
}

File: backend/lib/insurance_matching.py
from lib.finance_config import INSURANCE_EXAMPLES_PER_GAP_TYPE


# Closest sum-assured match to the computed gap — never "largest" or
# "cheapest" — to avoid the selection itself acting as a ranking signal
# (see .agents/context/subsystem-notes.md).
def select_insurance_examples(plan_type: str, gap_amount: float, plans: list[dict], count: int | None = None) -> list[dict]:
    if count is None:
        count = INSURANCE_EXAMPLES_PER_GAP_TYPE
    matching = [p for p in plans if p["plan_type"] == plan_type]
    matching.sort(key=lambda p: abs(p["sum_assured_max"] - gap_amount))
    return matching[:count]
